File: ios/Plugin/ThunderBgServicePlugin.swift
import Foundation
import Capacitor
import UserNotifications

/**
 * Plugin principal pour ThunderBgService sur iOS
 * Implémente les mêmes fonctionnalités que la version Android
 */
@objc(ThunderBgServicePlugin)
public class ThunderBgServicePlugin: CAPPlugin {
    
    private let notificationHelper = NotificationHelper.shared
    private let locationHelper = LocationHelper.shared
    private let taskManager = BackgroundTaskManager.shared
    private let geofenceHelper = GeofenceHelper.shared
    
    // Stockage des résultats de tâches
    private var taskResults: [String: Any] = [:]
    private var taskEventListeners: [String: (Any) -> Void] = [:]
    
    /**
     * Démarre le service foreground avec notification
     */
    @objc func start(_ call: CAPPluginCall) {
        guard let title = call.getString("notificationTitle") else {
            call.reject("notificationTitle is required")
            return
        }
        
        let subtitle = call.getString("notificationSubtitle")
        let enableLocation = call.getBool("enableLocation", false)
        let soundsEnabled = call.getBool("soundsEnabled", false)
        let customLayout = call.getString("customLayout")
        
        // Extraire viewData et buttons depuis les options
        var viewData: [String: String]?
        if let viewDataObj = call.getObject("viewData") {
            viewData = viewDataObj.compactMapValues { $0 as? String }
        }
        
        var buttons: [[String: Any]]?
        if let buttonsArray = call.getArray("buttons") {
            buttons = buttonsArray as? [[String: Any]]
        }
        
        // Démarrer la notification
        notificationHelper.showNotification(
            title: title,
            subtitle: subtitle,
            customLayout: customLayout,
            viewData: viewData,
            soundsEnabled: soundsEnabled
        )
        
        // Configurer les actions de notification (boutons)
        if let buttons = buttons {
            notificationHelper.configureNotificationActions(buttons: buttons)
        }
        
        // Démarrer la localisation si demandée
        if enableLocation {
            locationHelper.startLocationUpdates { [weak self] location in
                // Émettre un événement de localisation
                self?.notifyListeners("locationUpdate", data: [
                    "latitude": location.coordinate.latitude,
                    "longitude": location.coordinate.longitude,
                    "timestamp": Int64(location.timestamp.timeIntervalSince1970 * 1000)
                ])
            }
        }
        
        call.resolve(["started": true])
    }
    
    /**
     * Met à jour la notification
     */
    @objc func update(_ call: CAPPluginCall) {
        let title = call.getString("notificationTitle")
        let subtitle = call.getString("notificationSubtitle")
        
        // Extraire viewData
        var viewData: [String: String]?
        if let viewDataObj = call.getObject("viewData") {
            viewData = viewDataObj.compactMapValues { $0 as? String }
        }
        
        // Extraire buttons
        var buttons: [[String: Any]]?
        if let buttonsArray = call.getArray("buttons") {
            buttons = buttonsArray as? [[String: Any]]
        }
        
        notificationHelper.updateNotification(
            title: title,
            subtitle: subtitle,
            viewData: viewData
        )
        
        if let buttons = buttons {
            notificationHelper.configureNotificationActions(buttons: buttons)
        }
        
        call.resolve(["updated": true])
    }
    
    /**
     * Arrête le service
     */
    @objc func stop(_ call: CAPPluginCall) {
        notificationHelper.removeNotification()
        locationHelper.stopLocationUpdates()
        
        // Arrêter toutes les tâches
        let taskIds = taskManager.getRegisteredTasks()
        for taskId in taskIds {
            taskManager.unregisterTask(taskId: taskId)
        }
        
        // Supprimer toutes les géofences
        geofenceHelper.removeAllGeofences()
        
        call.resolve(["stopped": true])
    }
    
    /**
     * Enregistre une tâche en arrière-plan
     * Note: Sur iOS, les tâches sont limitées par le système
     */
    @objc func registerTask(_ call: CAPPluginCall) {
        guard let taskId = call.getString("taskId") else {
            call.reject("taskId is required")
            return
        }
        
        guard let intervalMs = call.getInt("intervalMs"), intervalMs >= 1000 else {
            call.reject("intervalMs must be >= 1000")
            return
        }
        
        // Sur iOS, on ne peut pas exécuter du code Swift arbitraire en arrière-plan
        // On simule avec un handler qui sera appelé par BGTaskScheduler
        let interval = TimeInterval(intervalMs) / 1000.0
        
        taskManager.registerTask(taskId: taskId, interval: interval) { [weak self] in
            // Exécuter la tâche
            self?.executeTask(taskId: taskId)
        }
        
        call.resolve(["registered": true])
    }
    
    /**
     * Désenregistre une tâche
     */
    @objc func unregisterTask(_ call: CAPPluginCall) {
        guard let taskId = call.getString("taskId") else {
            call.reject("taskId is required")
            return
        }
        
        taskManager.unregisterTask(taskId: taskId)
        taskResults.removeValue(forKey: taskId)
        
        call.resolve(["unregistered": true])
    }
    
    /**
     * Récupère le résultat d'une tâche
     */
    @objc func getTaskResult(_ call: CAPPluginCall) {
        guard let taskId = call.getString("taskId") else {
            call.reject("taskId is required")
            return
        }
        
        let result = taskResults[taskId]
        call.resolve(["result": result ?? NSNull()])
    }
    
    /**
     * Ajoute un listener pour les événements de tâches
     */
    @objc func addListener(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName"),
              eventName == "taskEvent" else {
            call.reject("Invalid event name")
            return
        }
        
        // Stocker le listener (sera appelé via notifyListeners)
        call.resolve(["remove": true])
    }
    
    /**
     * Supprime tous les listeners
     */
    @objc func removeAllListeners(_ call: CAPPluginCall) {
        taskEventListeners.removeAll()
        call.resolve()
    }
    
    /**
     * Exécute une tâche (méthode interne)
     */
    private func executeTask(taskId: String) {
        // Créer un résultat de tâche
        let result: [String: Any] = [
            "taskId": taskId,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
            "data": "Task executed"
        ]
        
        taskResults[taskId] = result
        
        // Émettre un événement si des listeners sont actifs
        notifyListeners("taskEvent", data: [
            "taskId": taskId,
            "data": result["data"] ?? "",
            "timestamp": result["timestamp"] ?? 0
        ])
        
        print("ThunderBG: Task executed: \(taskId)")
    }
    
    /**
     * Ajoute une géofence
     */
    @objc func addGeofence(_ call: CAPPluginCall) {
        guard let id = call.getString("id"),
              let latitude = call.getDouble("latitude"),
              let longitude = call.getDouble("longitude") else {
            call.reject("id, latitude, and longitude are required")
            return
        }
        
        let radius = call.getDouble("radius") ?? 100.0
        let onEnter = call.getString("onEnter")
        let onExit = call.getString("onExit")
        
        // Extraire les extras
        var extras: [String: String]?
        if let extrasObj = call.getObject("extras") {
            extras = extrasObj.compactMapValues { $0 as? String }
        }
        
        let config = GeofenceHelper.GeofenceConfig(
            id: id,
            latitude: latitude,
            longitude: longitude,
            radius: radius,
            onEnterAction: onEnter,
            onExitAction: onExit,
            extras: extras
        )
        
        geofenceHelper.addGeofence(config: config)
        
        call.resolve(["added": true])
    }
    
    /**
     * Supprime une géofence
     */
    @objc func removeGeofence(_ call: CAPPluginCall) {
        guard let geofenceId = call.getString("geofenceId") else {
            call.reject("geofenceId is required")
            return
        }
        
        geofenceHelper.removeGeofence(id: geofenceId)
        
        call.resolve(["removed": true])
    }
    
    /**
     * Supprime toutes les géofences
     */
    @objc func removeAllGeofences(_ call: CAPPluginCall) {
        geofenceHelper.removeAllGeofences()
        
        call.resolve(["removed": true])
    }
}

File: android/src/main/java/com/ahmedmili/thunderbgservice/core/ThunderBgServicePlugin.java
package com.ahmedmili.thunderbgservice.core;

import static com.ahmedmili.thunderbgservice.core.FgConstants.*;

import android.content.Intent;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.ahmedmili.thunderbgservice.tasks.TaskResultStorage;
import com.ahmedmili.thunderbgservice.geofencing.GeofenceManager;

@CapacitorPlugin(name = "ThunderBgService")
public class ThunderBgServicePlugin extends Plugin {
    private static ThunderBgServicePlugin instance;
    private GeofenceManager geofenceManager;
    
    @Override
    public void load() {
        super.load();
        instance = this;
        geofenceManager = new GeofenceManager(getContext());
    }
    
    public static ThunderBgServicePlugin getInstance() {
        return instance;
    }
    
    /**
     * Méthode publique pour émettre des événements depuis les classes externes.
     * Permet à TaskEventEmitter d'émettre des événements vers JS.
     */
    public void emitTaskEvent(String eventName, JSObject data) {
        notifyListeners(eventName, data);
    }
    
    @PluginMethod
    public void start(PluginCall call) {
        // S'assurer que l'instance est stockée
        if (instance == null) {
            instance = this;
        }
        String title = call.getString("notificationTitle", null);
        String subtitle = call.getString("notificationSubtitle", null);
        boolean enableLocation = call.getBoolean("enableLocation", true);
        boolean soundsEnabled = call.getBoolean("soundsEnabled", false);
        Intent extras = new Intent();
        if (title != null) extras.putExtra(EXTRA_TITLE, title);
        if (subtitle != null) extras.putExtra(EXTRA_SUBTITLE, subtitle);
        extras.putExtra(EXTRA_ENABLE_LOCATION, enableLocation);
        extras.putExtra(EXTRA_SOUNDS, soundsEnabled);
        // Optional custom layout from JS
        String customLayout = call.getString("customLayout", null);
        String titleViewId = call.getString("titleViewId", null);
        String subtitleViewId = call.getString("subtitleViewId", null);
        String timerViewId = call.getString("timerViewId", null);
        if (customLayout != null) extras.putExtra(EXTRA_CUSTOM_LAYOUT, customLayout);
        if (titleViewId != null) extras.putExtra(EXTRA_TITLE_VIEW_ID, titleViewId);
        if (subtitleViewId != null) extras.putExtra(EXTRA_SUBTITLE_VIEW_ID, subtitleViewId);
        if (timerViewId != null) extras.putExtra(EXTRA_TIMER_VIEW_ID, timerViewId);
        // Dynamic bindings (optional): viewData (object) and buttons (array)
        try {
            com.getcapacitor.JSObject viewData = call.getObject("viewData", null);
            if (viewData != null) {
                extras.putExtra(EXTRA_VIEW_DATA_JSON, viewData.toString());
            }
            com.getcapacitor.JSArray buttons = call.getArray("buttons", null);
            if (buttons != null) {
                extras.putExtra(EXTRA_BUTTONS_JSON, buttons.toString());
            }
        } catch (Exception ignored) {}
        ForegroundTaskService.startAction(getContext(), ACTION_START, extras);
        JSObject ret = new JSObject(); ret.put("started", true); call.resolve(ret);
    }

    @PluginMethod
    public void stop(PluginCall call) {
        ForegroundTaskService.startAction(getContext(), ACTION_STOP, null);
        JSObject ret = new JSObject(); ret.put("stopped", true); call.resolve(ret);
    }

    @PluginMethod
    public void update(PluginCall call) {
        Intent extras = new Intent();
        if (call.hasOption("notificationTitle")) { String v = call.getString("notificationTitle"); if (v != null) extras.putExtra(EXTRA_TITLE, v); }
        if (call.hasOption("notificationSubtitle")) { String v = call.getString("notificationSubtitle"); if (v != null) extras.putExtra(EXTRA_SUBTITLE, v); }
        // Optional: change layout dynamically
        String customLayout = call.getString("customLayout", null);
        String titleViewId = call.getString("titleViewId", null);
        String subtitleViewId = call.getString("subtitleViewId", null);
        String timerViewId = call.getString("timerViewId", null);
        if (customLayout != null) extras.putExtra(EXTRA_CUSTOM_LAYOUT, customLayout);
        if (titleViewId != null) extras.putExtra(EXTRA_TITLE_VIEW_ID, titleViewId);
        if (subtitleViewId != null) extras.putExtra(EXTRA_SUBTITLE_VIEW_ID, subtitleViewId);
        if (timerViewId != null) extras.putExtra(EXTRA_TIMER_VIEW_ID, timerViewId);
        // Dynamic bindings (optional)
        try {
            com.getcapacitor.JSObject viewData = call.getObject("viewData", null);
            if (viewData != null) {
                extras.putExtra(EXTRA_VIEW_DATA_JSON, viewData.toString());
            }
            com.getcapacitor.JSArray buttons = call.getArray("buttons", null);
            if (buttons != null) {
                extras.putExtra(EXTRA_BUTTONS_JSON, buttons.toString());
            }
        } catch (Exception ignored) {}
        ForegroundTaskService.startAction(getContext(), ACTION_UPDATE, extras);
        JSObject ret = new JSObject(); ret.put("updated", true); call.resolve(ret);
    }

    @PluginMethod
    public void registerTask(PluginCall call) {
        String taskId = call.getString("taskId", "");
        String taskClass = call.getString("taskClass", "");
        long intervalMs = call.getLong("intervalMs", 5000L);
        
        if (taskId.isEmpty() || taskClass.isEmpty() || intervalMs < 1000) {
            call.reject("Invalid parameters: taskId, taskClass required, intervalMs >= 1000");
            return;
        }
        
        Intent extras = new Intent();
        extras.putExtra(EXTRA_TASK_ID, taskId);
        extras.putExtra(EXTRA_TASK_CLASS, taskClass);
        extras.putExtra(EXTRA_TASK_INTERVAL, intervalMs);
        ForegroundTaskService.startAction(getContext(), ACTION_REGISTER_TASK, extras);
        
        JSObject ret = new JSObject();
        ret.put("registered", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void unregisterTask(PluginCall call) {
        String taskId = call.getString("taskId", "");
        if (taskId.isEmpty()) {
            call.reject("taskId is required");
            return;
        }
        
        Intent extras = new Intent();
        extras.putExtra(EXTRA_TASK_ID, taskId);
        ForegroundTaskService.startAction(getContext(), ACTION_UNREGISTER_TASK, extras);
        
        JSObject ret = new JSObject();
        ret.put("unregistered", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void getTaskResult(PluginCall call) {
        String taskId = call.getString("taskId", "");
        if (taskId.isEmpty()) {
            call.reject("taskId is required");
            return;
        }
        
        try {
            org.json.JSONObject result = TaskResultStorage.getResult(getContext(), taskId);
            JSObject ret = new JSObject();
            if (result != null) {
                // Convertir JSONObject en JSObject
                JSObject resultObj = new JSObject();
                java.util.Iterator<String> keys = result.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    Object value = result.get(key);
                    resultObj.put(key, value);
                }
                ret.put("result", resultObj);
            } else {
                ret.put("result", null);
            }
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Error getting task result: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void addGeofence(PluginCall call) {
        try {
            String id = call.getString("id", "");
            double latitude = call.getDouble("latitude", 0.0);
            double longitude = call.getDouble("longitude", 0.0);
            double radius = call.getDouble("radius", 100.0);
            String onEnter = call.getString("onEnter", null);
            String onExit = call.getString("onExit", null);
            
            if (id.isEmpty()) {
                call.reject("id is required");
                return;
            }
            
            GeofenceManager.GeofenceConfig config = new GeofenceManager.GeofenceConfig(
                id, latitude, longitude, (float) radius
            );
            config.onEnterAction = onEnter;
            config.onExitAction = onExit;
            
            // Extraire les extras
            try {
                JSObject extrasObj = call.getObject("extras", null);
                if (extrasObj != null) {
                    java.util.Iterator<String> keys = extrasObj.keys();
                    while (keys.hasNext()) {
                        String key = keys.next();
                        config.extras.put(key, extrasObj.getString(key));
                    }
                }
            } catch (Exception ignored) {}
            
            geofenceManager.addGeofence(config);
            
            JSObject ret = new JSObject();
            ret.put("added", true);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Error adding geofence: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void removeGeofence(PluginCall call) {
        String geofenceId = call.getString("geofenceId", "");
        if (geofenceId.isEmpty()) {
            call.reject("geofenceId is required");
            return;
        }
        
        geofenceManager.removeGeofence(geofenceId);
        
        JSObject ret = new JSObject();
        ret.put("removed", true);
        call.resolve(ret);
    }
    
    @PluginMethod
    public void removeAllGeofences(PluginCall call) {
        geofenceManager.removeAllGeofences();
        
        JSObject ret = new JSObject();
        ret.put("removed", true);
        call.resolve(ret);
    }
}

